File: examples/taskstore/go/api/result.go
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package api

import "fmt"

// Result is what the server answers with — the `sum` in cases/result.yaml.
//
// The arms are named for the outcome rather than the operation, because several
// operations share one: create, read and update all answer Found.
type Result interface{ isResult() }

type (
	Accepted struct{} // arity 0 — a delete that worked, nothing to send back
	Found    Task     // arity N — the task as it now stands
	Listing  TaskPage // arity N — a page of tasks plus the total
	Failed   APIError // arity N — the request did not work
)

func (Accepted) isResult() {}
func (Found) isResult()    {}
func (Listing) isResult()  {}
func (Failed) isResult()   {}

// Tag ordinals for Result, in the declaration order of cases/result.yaml.
const (
	resultAccepted uint8 = iota
	resultFound
	resultListing
	resultFailed
)

func appendResult(buf []byte, r Result) ([]byte, error) {
	switch v := r.(type) {
	case Accepted:
		return append(buf, resultAccepted), nil
	case Found:
		return appendTask(append(buf, resultFound), Task(v))
	case Listing:
		return appendPage(append(buf, resultListing), TaskPage(v))
	case Failed:
		return appendAPIError(append(buf, resultFailed), APIError(v))
	default:
		return nil, fmt.Errorf("taskstore: unhandled result %T", r)
	}
}

func readResult(b []byte) (Result, []byte, error) {
	tag, b, err := readU8(b)
	if err != nil {
		return nil, b, err
	}
	switch tag {
	case resultAccepted:
		return Accepted{}, b, nil
	case resultFound:
		t, b, err := readTask(b)
		return Found(t), b, err
	case resultListing:
		p, b, err := readPage(b)
		return Listing(p), b, err
	case resultFailed:
		e, b, err := readAPIError(b)
		return Failed(e), b, err
	default:
		return nil, b, fmt.Errorf("taskstore: unknown result tag %d", tag)
	}
}

File: examples/taskstore/rust/src/result.rs
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! What the server answers with: the `sum` in cases/result.yaml.

use serify::SerifyModel;

use crate::model::{ApiError, Task, TaskPage};
use crate::wire::Reader;

/// The arms are named for the outcome rather than the operation, because
/// several operations share one: create, read and update all answer `Found`.
///
/// The enum is `ApiResult` rather than `Result` only to leave that name to
/// `std`; the schema tags come from the *variant* names, so the type's own name
/// never reaches the wire.
#[derive(SerifyModel, Clone, Debug, PartialEq)]
pub enum ApiResult {
    Accepted,          // a delete that worked: nothing to send back
    Found(Task),       // the task as it now stands
    Listing(TaskPage), // a page of tasks plus the total
    Failed(ApiError),  // the request did not work
}

impl ApiResult {
    pub fn put(&self, buf: &mut Vec<u8>) -> Result<(), String> {
        match self {
            ApiResult::Accepted => buf.push(0),
            ApiResult::Found(t) => {
                buf.push(1);
                t.put(buf)?;
            }
            ApiResult::Listing(p) => {
                buf.push(2);
                p.put(buf)?;
            }
            ApiResult::Failed(e) => {
                buf.push(3);
                e.put(buf)?;
            }
        }
        Ok(())
    }

    pub fn read(r: &mut Reader) -> Result<Self, String> {
        match r.u8()? {
            0 => Ok(ApiResult::Accepted),
            1 => Ok(ApiResult::Found(Task::read(r)?)),
            2 => Ok(ApiResult::Listing(TaskPage::read(r)?)),
            3 => Ok(ApiResult::Failed(ApiError::read(r)?)),
            tag => Err(format!("unknown result tag {tag}")),
        }
    }
}
